Source/IGListGridCollectionViewLayout.m

```diff
@@ -0,0 +1,422 @@
+
+- (void)reloadLayout
+{
+    [self.lineCache removeAllObjects];
+    [self.lineForItem removeAllObjects];
+    
+    // Init first line and add to lineCache
+    CGRect frame = CGRectMake(0, 0, self.contentWidth, 0);
+    _IGListGridLayoutLine *firstLine = [[_IGListGridLayoutLine alloc] initWithMinimumInteritemSpacing:self.minimumInteritemSpacing
+                                                                                    headIndex:0
+                                                                                        frame:frame
+                                                                              scrollDirection:self.scrollDirection];
+    [self.lineCache addObject:firstLine];
+    
+    for (NSInteger i = 0; i < self.collectionView.numberOfSections; i++) {
+        NSIndexPath *indexPath = [NSIndexPath indexPathForItem:0 inSection:i];
+        id<UICollectionViewDelegateFlowLayout> delegate = (id<UICollectionViewDelegateFlowLayout>) self.collectionView.delegate;
+        CGSize itemSize = [delegate collectionView:self.collectionView layout:self sizeForItemAtIndexPath:indexPath];
+        _IGListGridLayoutLine *lastLine = [self.lineCache lastObject];
+        if (![lastLine addItemToTailWithSize:itemSize]) {
+            // Not enough space for the last line
+            CGFloat y = lastLine.frame.origin.y + lastLine.frame.size.height + self.minimumLineSpacing;
+            frame = CGRectMake(0, y, self.contentWidth, 0);
+            _IGListGridLayoutLine *newLine = [[_IGListGridLayoutLine alloc] initWithMinimumInteritemSpacing:self.minimumInteritemSpacing
+                                                                                            headIndex:i
+                                                                                                frame:frame
+                                                                                      scrollDirection:self.scrollDirection];
+            [self.lineCache addObject:newLine];
+            [newLine addItemToTailWithSize:itemSize];
+        }
+        [self.lineForItem addObject:[NSNumber numberWithInteger:(self.lineCache.count - 1)]];
+    }
+}
+
+- (void)reloadLayoutWithConstantItemSize:(CGSize)itemSize
+{
+    self.itemPerLine = (int) ((self.contentWidth + self.minimumInteritemSpacing)
+                               / (itemSize.width + self.minimumInteritemSpacing));
+    self.lineNumber = (self.collectionView.numberOfSections + self.itemPerLine - 1) / self.itemPerLine;
+    self.interitemSpacing = (self.contentWidth - (self.itemSize.width * self.itemPerLine)) / (self.itemPerLine - 1);
+}
+
+@end
+
+#pragma mark - _IGListGridLayoutLine Implementation
+
+@implementation _IGListGridLayoutLine
+
+- (id)initWithMinimumInteritemSpacing:(CGFloat)spacing
+                            headIndex:(NSInteger)headIndex
+                                frame:(CGRect)frame
+                      scrollDirection:(UICollectionViewScrollDirection)direction
```

Comment:
Should we assert any state of these params? e.g. things that should be `> 0`, etc